syftlib/src/main/java/org/openmined/syft/SignallingClient.kt

```diff
@@ -0,0 +1,141 @@
+                try {
+                    createSocket()
+                    while (socket.isConnected) {
+                        emitter.onNext(reader.readLine())
+                    }
+                    emitter.onComplete()
+                } catch (e: Exception) {
+                    Log.e(TAG, "Socket TimeOut", e)
+                    stop()
+                    emitter.onError(e)
+                }
+            }, BackpressureStrategy.BUFFER
+        )
+
+        disposable = invokeSocketObserver()
+        sendingThread = runKeepAliveSchedule()
+    }
+
+    /**
+     * Send the data over the Socket connection to PyGrid
+     */
+    fun send(type: String, data: JsonObject) {
+        if (socket.isConnected) {
+            sendingThread.cancel(false)
+            data.addProperty("workerId", workerId)
+            val message = JsonObject().apply {
+                this.addProperty("type", type)
+                this.add("data", data)
+            }
+            writer.write(message.toString())
+            writer.flush()
+            Log.d(TAG, "message sent :$message")
+            sendingThread = runKeepAliveSchedule()
+        } else {
+            disposable.dispose()
+            invokeSocketObserver()
+        }
+
+    }
+
+    private fun runKeepAliveSchedule(): ScheduledFuture<*> {
+        return keepSendingScheduler.scheduleAtFixedRate(
+            {
+                if (socket.isConnected)
+                    send(SOCKET_PING, JsonObject())
+            },
+            keepAliveTimeout.toLong(),
+            keepAliveTimeout.toLong(),
+            TimeUnit.MILLISECONDS
+        )
+    }
+
+    /**
+     * @return void but sets up SignallingClient class properties
+     */
+    private fun createSocket() {
+        Log.d(TAG, "Opening Socket Connection")
+        socket = Socket(url, port)
+        Log.d(TAG, "Socket Connected")
+        socket.keepAlive = true
+        socket.soTimeout = keepAliveTimeout
+        writer = socket.getOutputStream().bufferedWriter(Charset.defaultCharset())
+        reader = socket.getInputStream().bufferedReader(Charset.defaultCharset())
+        signallingInterface.onOpen()
+    }
+
+    /**
+     * Whenever a subscriber is subscribed to Observable, a new invocation happens
+     * initiating socket connection and setting up socket reader and writer
+     *
+     * This function is also called when connection is no more `Connected`
+     * thereby re-establishing connection to PyGrid
+     */
+    private fun invokeSocketObserver(): Disposable {
+        return responseObservable
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribeOn(Schedulers.io())
+            .subscribe(
+                { signallingInterface.onMessage(it) },
+                { Log.e(TAG, "we have a socket error!", it) },
+                {
+                    Log.d(TAG, "Socket disconnected")
+                    invokeSocketObserver()
```

Comment:
I'm not sure what is the behaviour of this function invoking itself. 
If there is no connection to PyGrid (`socket.isConnected` will return false), will this function be creating a socket continuously?

Reply:
If socket wasn't connected probably it had died down. I actually wanted to retry connecting to socket. Hmm maybe I should call socket.connect in that case. We wont be creating new socket objects atleast that way